File: haskell/word-count/src/WordCount.hs
module WordCount (wordCount) where

import Control.Applicative (liftA2)
import Data.Char (isAlphaNum, isSpace, toLower)
import Data.List (dropWhileEnd, foldl')
import Data.List.Split (wordsBy)
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map

-- | Given a string, returns a string with each character converted
-- to lowercase. Also found in "Distribution.Simple.Utils".
lowercase :: String -> String
lowercase = map toLower

-- | Given a string, returns a list of strings by splitting on (and dropping)
-- non-alphanumeric characters.
words' :: String -> [String]
words' =
  map unQuote
    . wordsBy (not . (isAlphaNum <||> isQuote))

-- | Given a string, returns a Map from word to frequency for each word.
wordCount :: String -> Map String Int
wordCount = foldl' go Map.empty . map lowercase . words'
  where
    go seen word = Map.insertWith (+) word 1 seen

unQuote :: String -> String
unQuote = dropWhileEnd isQuote . dropWhile isQuote

isQuote :: Char -> Bool
isQuote '\'' = True
isQuote _ = False

(<||>) :: (Applicative f) => f Bool -> f Bool -> f Bool
(<||>) = liftA2 (||)

infixr 2 <||>

File: go/word-count/word_count.go
// Package wordcount implements the Word Count exercism from Exercism.
package wordcount

import (
	"regexp"
	"strings"
)

// Frequency is a map from word to the number of occurrences.
type Frequency map[string]int

// WordCount computes the frequency of words in a given a phrase.
func WordCount(phrase string) Frequency {
	nonAlphanumeric := regexp.MustCompile(`[^[:alpha:][:digit:]']+`)
	words := nonAlphanumeric.Split(strings.ToLower(phrase), -1)
	frequency := Frequency{}
	for _, word := range words {
		normalizedWord := strings.Trim(strings.ToLower(word), `'`)
		if len(normalizedWord) == 0 {
			continue
		}
		frequency[normalizedWord]++
	}
	return frequency
}
